_posts/2023-03-11-Adding-a-Timeout-feature-to-Rustgem.md

```diff
@@ -0,0 +1,288 @@
+        // After this, all fields are initialized.
+        unsafe {
+            addr_of_mut!((*t).inner).write(inner);
+            bindings::timer_setup(addr_of_mut!((*t).timer), Some(timer_callback::<T, D>), 0)
+        };
+
+        Self(t)
+    }
+}
+```
+
+And them, the `container_of` macro is now working! Now, I could setup a Timer
+for each fence and keep the fence inside the timer. Finally, I could use the
+fence inside the timer to signal it when it was not signaled by the user for
+more than 10 seconds.
+
+```rust
+impl TimerOps for VgemFenceOps {
+    type Inner = UniqueFence<Self>;
+
+    fn timer_callback(timer: &UniqueTimer<Self, UniqueFence<Self>>) {
+        let _ = timer.inner().signal();
+    }
+}
+```
+
+So, I tested the driver with IGT using the `vgem_slow` test and it was now
+passing! All IGT tests were passing and it looked like the driver was
+practically completed (a part from some FIXME problems). But, let's see if this
+abstraction is really safe...
+
+# Second Attempt: using a Timer abstraction
+---
+
+First, let's inspect the `struct timer_list` in the C code.
+
+```c
+struct timer_list {
+    struct hlist_node   entry;
+    unsigned long       expires;
+    void            (*function)(struct timer_list *);
+    u32         flags;
+};
+```
+
+So, by looking at this struct, we can see a problem in my abstraction: a timer
+can point to a timer through a list. If you are not familiar with Rust, this can
+seem normal, but self-referential types can lead to undefined behavior (UB).
+
+Let's say we have an example type with two fields: `u32` and a pointer to this
+`u32` value. Initially, everything looks fine, the pointer field points to the
+value field in memory address A, which contains a valid `u32`, and all pointers
+are **valid**. But Rust has the freedom to move values around memory. For
+example, if we pass this struct into another function, it might get moved to a
+different memory address. So, the once valid pointer is no longer valid, because
+when we move the struct, the struct's fields change their address, but not their
+value. So, the pointer fields still point to the memory address A, although the
+value field is located at the memory address B now. This is really bad and can
+lead to UB.
+
+The solution is to make `timer_list` implement the `!Unpin` trait. This means
+that to use this type safely, we can't use regular pointers for self-reference.
+Instead, we use special pointers that "pin" their values into place, ensuring
+they can't be moved.
+
+Still looking at the `struct timer_list`, it is possible to notice that a timer
+can queue itself in the timer function. This functionality is not covered by my
+current abstraction.
+
+Moreover, I was using jiffies to modify the timeout duration and I was adding a
+`Duration` to the jiffies. This is problematic, because this can cause a data
+race. So, the reading of jiffies should be atomic and adding a Duration to
+jiffies should be a atomic operation.
```

Comment:
```suggestion
`Duration` to the jiffies. This is problematic, because it can cause data
races. Reading jiffies and adding a duration to them should be an atomic operation.
```